### PR TITLE
feat(otlp): experimental otlp support

### PR DESCRIPTION
**Pull request by petethepig**
* very much a draft at this point.
* requires a copy of https://github.com/open-telemetry/opentelemetry-proto-profile to be adjacent to the directory
* works with https://github.com/grafana/pyroscope-golang/pull/36 and https://github.com/petethepig/opentelemetry-collector/pull/1

### CODE REVIEW

File: pkg/api/api.go [hunk 1]
```diff
@@ -197,13 +199,19 @@ func (a *API) RegisterOverridesExporter(oe *exporter.OverridesExporter) {
...
+	otlpHandler := otlp.NewOTLPIngestHandler(d, a.logger)
 	a.RegisterRoute("/ingest", pyroscopeHandler, true, true, "POST")
 	a.RegisterRoute("/pyroscope/ingest", pyroscopeHandler, true, true, "POST")
 	pushv1connect.RegisterPusherServiceHandler(a.server.HTTP, d, a.grpcAuthMiddleware)
 	a.RegisterRoute("/distributor/ring", d, false, true, "GET", "POST")
 	a.indexPage.AddLinks(defaultWeight, "Distributor", []IndexPageLink{
 		{Desc: "Ring status", Path: "/distributor/ring"},
 	})
+
+	pprofileotlp.RegisterGRPCServer(a.server.GRPC, otlpHandler)
```
Comment (cyriltovena) [inline pkg/api/api.go:211]: Just so you know I don't think we expose GRPC endpoint anywhere in our infra nor in helm.

Comment (petethepig) [reply pkg/api/api.go:211]: @cyriltovena yeah we might need a separate grpc server

File: go.mod [hunk 2]
```diff
@@ -283,3 +283,10 @@ replace (
 	// + https://github.com/go-yaml/yaml/pull/876
 	gopkg.in/yaml.v3 => github.com/colega/go-yaml-yaml v0.0.0-20220720105220-255a8d16d094
 )
+
+require (
+	github.com/jzelinskie/must v0.0.1
+	go.opentelemetry.io/collector/pdata v1.0.0-rcv0012
+)
+
+replace go.opentelemetry.io/collector/pdata => ../opentelemetry-collector/pdata
```
Comment (petethepig) [inline go.mod:292]: to be clear, we won't have this dependency once this is ready to merge

File: pkg/ingester/otlp/ingest_handler.go [hunk 3]
```diff
@@ -65,41 +67,46 @@ func (h *ingestHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
...
+		fmt.Println("service.name: ", serviceName)
+
+		// Start with default labels
+		labels := getDefaultLabels(serviceName)
+
+		// Track processed attribute keys to avoid duplicates across levels
+		processedKeys := make(map[string]bool)
+
+		// Add resource attributes
+		labels = appendAttributesUnique(labels, rp.Resource.GetAttributes(), processedKeys)
```
Comment (marcsanmi) [inline pkg/ingester/otlp/ingest_handler.go:92]: Right now we're adding all attributes as labels... do we want to set some limit?

File: pkg/ingester/otlp/ingest_handler.go [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+
+		labelsDst := []*typesv1.LabelPair{}
+		// TODO(@petethepig): make labels work
+		labelsDst = append(labelsDst, &typesv1.LabelPair{
+			Name:  "__name__",
+			Value: "process_cpu",
+		})
+		labelsDst = append(labelsDst, &typesv1.LabelPair{
+			Name:  "service_name",
+			Value: "otlp_test_app4",
```
[the remark this answers is not in the document]
Comment (marcsanmi) [reply pkg/ingester/otlp/ingest_handler.go]: Done in ff60e7c1a827a7f98b349adf626211800449fb09

File: pkg/ingester/otlp/ingest_handler.go [hunk 5]
```diff
@@ -65,41 +67,46 @@ func (h *ingestHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
...
-
-	// TODO(@petethepig): make it tenant-aware
-	ctx = tenant.InjectTenantID(ctx, tenant.DefaultTenantID)
-
-	h.log.Log("msg", "Export called")
+	// Extracts user ID from the request metadata and returns and injects the user ID in the context
+	_, ctx, err := user.ExtractFromGRPCRequest(ctx)
+	if err != nil {
+		level.Error(h.log).Log("msg", "failed to extract tenant ID from GRPC request", "err", err)
+		return &pprofileotlp.ExportProfilesServiceResponse{}, fmt.Errorf("failed to extract tenant ID from GRPC request: %w", err)
```
Comment (marcsanmi) [inline pkg/ingester/otlp/ingest_handler.go:74]: I assumed we want to fail want we're able to extract tenant id?

**Comment by fandreuz** [comment]
Hi @petethepig, thanks for working on this feature. I'm using Pyroscope to prototype OTLP support in Async-Profiler. I noticed this went a bit out of sync with the OTLP definition for `Profile` (open-telemetry/opentelemetry-proto#644, open-telemetry/opentelemetry-proto#650). 

Do you plan to keep this closely in sync with the protocol? Would you be open to reviewing a PR for this?

**Comment by korniltsev** [comment]
@fandreuz We will update the outdated otlp once we have some spare cycles. PR's are are very much welcome and appreciated, we'll be happy to review and mergge

**Pull request merged**